utils/vcsdetails.go

```diff
@@ -0,0 +1,120 @@
+package utils
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
+)
+
+type (
+	VcsCache struct {
+		// Key - Path to the .git directory.
+		// Value - Reference to a struct, storing the URL and revision.
+		vcsRootDir sync.Map
+		// Key - Path to a directory.
+		// Value - Reference to a struct, storing the URL and revision from the upstream .git. Can also include nil, if there's no upstream .git.
+		vcsDir sync.Map
+		// The current size of vcsDir
+		vcsDirSize *int32 // Size of vcs folders entries
+	}
+	vcsDetails struct {
+		url      string
+		revision string
+	}
+)
+
+const MAX_ENTRIES = 10000
+
+func NewVcsDetals() *VcsCache {
+	return &VcsCache{vcsRootDir: sync.Map{}, vcsDir: sync.Map{}, vcsDirSize: new(int32)}
+}
+
+func (this *VcsCache) incCacheSize(num int32) {
+	atomic.AddInt32(this.vcsDirSize, num)
+}
+
+func (this *VcsCache) getCacheSize() int32 {
+	return atomic.LoadInt32(this.vcsDirSize)
+}
+
+// Search for '.git' directory at 'path', incase there is one, extract the details and add a new entry to the cache(hash-map,key:path,value:pointer to the git details).
```

Comment:
cache(hash-map,key:path,value:pointer to the git details) --> This part is a but cryptic. Let's try to rephrase.

Reply:
`cache(key:path in the file system  ,value: git revision & url).`